lib/skate/detours/route_segments.ex

```diff
@@ -21,49 +21,73 @@ defmodule Skate.Detours.RouteSegments do
-          detour: [%Util.Location{latitude: 0, longitude: 1}, %Util.Location{latitude: 0, longitude: 2}],
-          after_detour: [%Util.Location{latitude: 0, longitude: 2}, %Util.Location{latitude: 0, longitude: 3}]
+          before_detour: [
+            %Util.Location{latitude: 0.0, longitude: 0.0},
+            %Util.Location{latitude: 0.0, longitude: 1.0},
+            %Util.Location{latitude: 0.0, longitude: 1.3}
+          ],
+          detour: [
+            %Util.Location{latitude: 0.0, longitude: 1.3},
+            %Util.Location{latitude: 0.0, longitude: 2.0},
+            %Util.Location{latitude: 0.0, longitude: 2.6}
+          ],
+          after_detour: [
+            %Util.Location{latitude: 0.0, longitude: 2.6},
+            %Util.Location{latitude: 0.0, longitude: 3.0}
+          ]
         }}
   """
   @spec route_segments(
           nonempty_list(Util.Location.From.t()),
           Util.Location.From.t(),
           Util.Location.From.t()
         ) :: {:ok, Result.t()} | :error
-  def route_segments(
-        [],
-        _connection_start,
-        _connection_end
-      ),
-      do: :error
+  def route_segments([], _connection_start, _connection_end), do: :error
+  def route_segments([_], _connection_start, _connection_end), do: :error
 
   def route_segments(
         shape,
         connection_start,
         connection_end
       ) do
-    {_, start_index} =
-      shape
-      |> Enum.with_index()
-      |> Enum.min_by(fn {point, _index} -> Util.Location.distance(point, connection_start) end)
+    {nearest_start_point, start_index} =
+      nearest_point_to_shape(shape, connection_start)
 
-    {_, end_index} =
-      shape
-      |> Enum.with_index()
-      |> Enum.min_by(fn {point, _index} -> Util.Location.distance(point, connection_end) end)
+    {nearest_end_point, end_index} =
+      nearest_point_to_shape(shape, connection_end)
 
     {:ok,
      %__MODULE__.Result{
-       before_detour: Enum.slice(shape, 0..start_index),
-       detour: Enum.slice(shape, start_index..end_index),
-       after_detour: Enum.slice(shape, end_index..-1)
+       before_detour: Enum.slice(shape, 0..start_index) ++ [nearest_start_point],
+       detour:
+         [nearest_start_point] ++
+           Enum.slice(shape, (start_index + 1)..end_index) ++ [nearest_end_point],
+       after_detour: [nearest_end_point] ++ Enum.slice(shape, (end_index + 1)..-1)
      }}
   end
+
+  defp nearest_point_to_shape(shape, point) do
+    {%{nearest_point: nearest_point}, index} =
+      shape
+      |> Enum.zip(Enum.drop(shape, 1))
+      |> Enum.map(fn segment -> Util.Location.nearest_point_to_segment(point, segment) end)
+      |> Enum.map(fn nearest_point ->
+        %{nearest_point: nearest_point, distance: Util.Location.distance(nearest_point, point)}
+      end)
+      |> Enum.with_index()
+      |> Enum.min_by(fn {%{distance: distance}, _} -> distance end)
```

Comment:
question(non-blocking): forgive me if I'm missing something, but do any consumers of this function use the `:distance` return value?

This may be worse for performance, but I was curious if the distance value ever needed to make it into the list at all.

```suggestion
      |> Enum.with_index()
      |> Enum.min_by(fn {%{nearest_point: nearest_point}, _} -> Util.Location.distance(nearest_point, point) end)

```

Reply:
Oh yeah that's a good call. I don't think it's worse for performance, because we have to compute the distance that number of times anyway. And if it turns out that we do wind up needing it, then we can start storing it then.

Reply:
[That's much nicer!](https://github.com/mbta/skate/pull/2506/commits/84638ddf08338df9b2a4cc5acc46b0cb3e6f8b6d)

Reply:
> I don't think it's worse for performance, because we have to compute the distance that number of times anyway.

Oh! I just read this wrong, I saw `min_by` and thought `sort` 😅 
Glad this turned out well!